File: Week_02/144/binary_tree_preorder_traversal3.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        res = []

        while root is not None:
            res.append(root.val)
            if root.left is None:
                root = root.right
            else:
                pre = root.left
                while pre.right is not None:
                    pre = pre.right
                pre.right = root.right
                root = root.left

        return res

    def test(self):
        n1 = TreeNode(1)
        n2 = TreeNode(2)
        n3 = TreeNode(3)

        n1.right = n2
        n2.left = n3

        ans = self.preorderTraversal(n1)
        assert ans == [1, 2, 3], f"ans: {ans}"
        print("well done")


if __name__ == "__main__":
    Solution().test()

File: Week_02/144/binary_tree_preorder_traversal3.go
package main

func preorderTraversal3(root *TreeNode) []int {
	res := make([]int, 0)

	for root != nil {
		res = append(res, root.Val)
		if root.Left == nil {
			root = root.Right
		} else {
			pre := root.Left

			for pre.Right != nil {
				pre = pre.Right
			}
			pre.Right = root.Right
			root = root.Left
		}
	}
	return res
}
